fuel-core/src/executor.rs

```diff
@@ -267,86 +283,54 @@ impl Executor {
                         AssetId::BASE,
                     )],
                 );
-                // TODO: Use a linked list instead of a vector to do it faster.
-                block.transactions.insert(0, mint.clone().into());
+                transactions.insert(0, mint.clone().into());
                 mint
             }
             ExecutionKind::Validation => {
-                let mint =
-                    if let Some(Transaction::Mint(mint)) = block.transactions.get(0) {
-                        mint.clone()
-                    } else {
-                        return Err(Error::CoinbaseIsNotFirstTransaction)
-                    };
+                let mint = if let Some(Transaction::Mint(mint)) = transactions.get(0) {
+                    mint.clone()
+                } else {
+                    return Err(Error::CoinbaseIsNotFirstTransaction)
+                };
                 self.check_coinbase(block_height as Word, mint, None)?
             }
         };
-
-        // Skip `coinbase` from execution.
-        let tx_iter = block.transactions.iter_mut().enumerate().skip(1);
+        // Skip the coinbase transaction.
+        let mut tx_index = 1;
 
         // Execute each transaction.
-        for (idx, tx) in tx_iter {
```

Comment:
why do we need to change this to a while loop with manually managed index incrementing? Is it just for skipping the first tx?

Reply:
It was done to be able to use `swap_remove` below 
<img width="398" alt="image" src="https://user-images.githubusercontent.com/18346821/199624659-04bf0005-eb84-49ef-9fbb-02bdcb0c8fd7.png">


With an iterator, it is not possible. Alternative to the `while` loop, I can transform transactions into `into_iterator`, filter them, and after collect them into vector and set to the block.

Reply:
You could use [retain_mut](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain_mut)

Reply:
Hmm I don't think we'd want to swap_remove, as wouldn't that change the sorted order of the transactions and potentially impact the validity of the block?

For example, let's say you have 3 transactions (tx1, tx2, tx3):
- let tx3 depend on the output of tx2
- let tx1 be a faulty tx that needs to be skipped

If tx1 is swap_removed, the vec of transactions will now be [tx3, tx2]. Worse, tx3 wont even be processed at all.

I made a simple playground to demonstrate how this can lead to pretty severely incorrect results on a larger set of txs where half of the txs are invalid: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50c1cc2edb68d931ebd99aa3e2ce58bc

I'd rather focus of the code be on correctness and clarity, vs optimizing array access and modification.

As a remedy, I'd prefer to lean more heavily on iterator style operations like [FilterMap](https://doc.rust-lang.org/std/iter/struct.FilterMap.html) as opposed to imperatively mucking with the memory.

Reply:
Rust is also specially tuned to speed up iterators vs loops. It allows the compiler to do more inlining and other optimizations that aren't possible with side-effectful imperative looping styles: https://doc.rust-lang.org/book/ch13-04-performance.html

Reply:
retain_mut also looks like a good option

Reply:
Also, I don't think we'd want to increment the index if the tx was skipped. As the index is used for txpointers and other indexes where we want to record the total location of a tx within a block. If there are skipped txs, these indexes will point to incorrect locations in the block.

Reply:
In while we don't increment it for skipped transacitons=)

Reply:
ah right - the `continue`

Reply:
The changes look better, although since the original bug wasn't caught in the original unit tests I think we should add new cases to cover the issue I uncovered above.

Reply:
Done